EIPS/eip-7775.md

```diff
@@ -12,62 +12,38 @@ created: 2024-09-30
 
 ## Specification
 
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
 The `BURN` opcode `(0xFC)` is introduced with the following behavior:
 
-1. Pops one 32-byte word from the top of the stack, treating these bytes as the `uint256` amount of native token to be burned.
+1. Pops one 32-byte word from the top of the stack, treating these bytes as the `uint256` amount of native ether to be burned.
 2. Retrieves the current address from the EVM execution context.
 3. Checks the balance of the current address.
 4. If the amount to be burned is greater than the balance of the current address, the opcode `MUST` revert.
 5. If the amount to be burned is 0, the execution `MUST NOT` revert.
-6. Subtracts the amount from the current address's native token balance.
+6. Subtracts the amount from the current address's native ether balance.
 
 When `BURN` is used in the context of a DELEGATECALL or CALLCODE, the contract whose balance is to be manipulated is the contract that issued the DELEGATECALL or CALLCODE instruction.
 
 When `BURN` is used in the context of a STATICCALL the call `MUST` revert.
 
-Example pseudocode for the `BURN` opcode:
-
-```python
-def op_burn(pc, interpreter, scope):
-    # Pop the value to be burned from the stack
-    value_to_burn = scope.stack.pop()
-    
-    # Retrieve the current address from the EVM execution context
-    current_address = scope.contract.address()
-    
-    # Check the balance of the current address
-    balance = interpreter.evm.state_db.get_balance(current_address)
-    
-    # If the value to be burned is greater than the balance, revert
-    if value_to_burn > balance:
-        return "ErrInsufficientBalance"
-    
-    # If the value to be burned is 0, do not revert
-    if value_to_burn == 0:
-        return None
-    
-    # Subtract the value from the current address's balance
-    interpreter.evm.state_db.sub_balance(current_address, value_to_burn)
-    return None
-```
-
-The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+When the `BURN` opcode is processed, 0 gas MUST be consumed.
```

Comment:
This is definitely too low. The reason is that this operation does a state change: it will change the balance of this contract. This is an expensive operation as it will hit the disk. In case of "just sending ether" (CALL opcode) at least 9100 gas has to be paid. The BURN operation here is simpler, because it will not try to invoke another contract, it will only change the balance of the current contract.

Reply:
Also note that simple operations like additions cost 3 gas, and this is an operation with much more complexity

Reply:
@jochem-brouwer what do you suggest, I was putting 0 as a placeholder for now? Is there a general philsophy for how gas costs are determined for new opcodes?

Reply:
I am not entirely aware how gas costs are decided on in the end, I think EIPs initially propose costs which are likely too high. These could then be adjusted downward later.

For instance see the gas spec of 1153 motivation:

> Gas cost for TSTORE is the same as a warm SSTORE of a dirty slot (i.e. original value is not new value and is not current value, currently 100 gas), and gas cost of TLOAD is the same as a hot SLOAD (value has been read before, currently 100 gas). Gas cost cannot be on par with memory access due to transient storage’s interactions with reverts.

( https://eips.ethereum.org/EIPS/eip-1153, specification section )

I would set 9100 gas here with the comment that this is the same as the (minimal) gas used for a CALL operation with value. Benchmarks/experiments could show that these costs are too high and that it is safe to set this to a lower number.

Reply:
@jochem-brouwer I added a simple algorithm, let me know what you think!

Reply:
kept it reasonable, and used SLOAD, SSTORE as a reference, since `BURN` would be effectively reading and writing 32 bytes.

Reply:
100 is too low. Please check a gas schedule for state-changing operations, such as `CALL` or `SSTORE`, a good gas overview can be found on https://evm.codes

![image](https://github.com/user-attachments/assets/b9eb15cb-4f04-4dc4-a942-4288f790a451)

See for instance SSTORE changing 10000 -> 500. This costs 2900 gas. One could argue this would be equal to a contract having 10000 balance and now burning 9500. (So 2900 is reasonable here).

Reply:
:+1: